File: tests/algorithms/x/Java/maths/numerical_analysis/simpson_rule.java
public class Main {
    static double result;

    static double f(double x) {
        return (x - 0.0) * (x - 0.0);
    }

    static double[] make_points(double a, double b, double h) {
        double[] points = ((double[])(new double[]{}));
        double x = a + h;
        while (x < (b - h)) {
            points = ((double[])(java.util.stream.DoubleStream.concat(java.util.Arrays.stream(points), java.util.stream.DoubleStream.of(x)).toArray()));
            x = x + h;
        }
        return points;
    }

    static double simpson_rule(double[] boundary, int steps) {
        if (steps <= 0) {
            throw new RuntimeException(String.valueOf("Number of steps must be greater than zero"));
        }
        double a = boundary[0];
        double b = boundary[1];
        double h = (b - a) / (((Number)(steps)).doubleValue());
        double[] pts = ((double[])(make_points(a, b, h)));
        double y = (h / 3.0) * f(a);
        int cnt = 2;
        int i = 0;
        while (i < pts.length) {
            double coeff = 4.0 - 2.0 * (((Number)((Math.floorMod(cnt, 2)))).doubleValue());
            y = y + (h / 3.0) * coeff * f(pts[i]);
            cnt = cnt + 1;
            i = i + 1;
        }
        y = y + (h / 3.0) * f(b);
        return y;
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            result = simpson_rule(((double[])(new double[]{0.0, 1.0})), 10);
            System.out.println(_p(result));
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static String _p(Object v) {
        if (v == null) return "<nil>";
        if (v.getClass().isArray()) {
            if (v instanceof int[]) return java.util.Arrays.toString((int[]) v);
            if (v instanceof long[]) return java.util.Arrays.toString((long[]) v);
            if (v instanceof double[]) return java.util.Arrays.toString((double[]) v);
            if (v instanceof boolean[]) return java.util.Arrays.toString((boolean[]) v);
            if (v instanceof byte[]) return java.util.Arrays.toString((byte[]) v);
            if (v instanceof char[]) return java.util.Arrays.toString((char[]) v);
            if (v instanceof short[]) return java.util.Arrays.toString((short[]) v);
            if (v instanceof float[]) return java.util.Arrays.toString((float[]) v);
            return java.util.Arrays.deepToString((Object[]) v);
        }
        return String.valueOf(v);
    }
}

File: tests/algorithms/x/Kotlin/maths/numerical_analysis/simpson_rule.kt
fun panic(msg: String): Nothing { throw RuntimeException(msg) }

fun _numToStr(v: Number): String {
    val d = v.toDouble()
    val i = d.toLong()
    return if (d == i.toDouble()) i.toString() else d.toString()
}

var _nowSeed = 0L
var _nowSeeded = false
fun _now(): Long {
    if (!_nowSeeded) {
        System.getenv("MOCHI_NOW_SEED")?.toLongOrNull()?.let {
            _nowSeed = it
            _nowSeeded = true
        }
    }
    return if (_nowSeeded) {
        _nowSeed = (_nowSeed * 1664525 + 1013904223) % 2147483647
        kotlin.math.abs(_nowSeed)
    } else {
        kotlin.math.abs(System.nanoTime())
    }
}

fun toJson(v: Any?): String = when (v) {
    null -> "null"
    is String -> "\"" + v.replace("\"", "\\\"") + "\""
    is Boolean, is Number -> v.toString()
    is Map<*, *> -> v.entries.joinToString(prefix = "{", postfix = "}") { toJson(it.key.toString()) + ":" + toJson(it.value) }
    is Iterable<*> -> v.joinToString(prefix = "[", postfix = "]") { toJson(it) }
    else -> toJson(v.toString())
}

var result: Double = simpson_rule(mutableListOf(0.0, 1.0), 10)
fun f(x: Double): Double {
    return (x - 0.0) * (x - 0.0)
}

fun make_points(a: Double, b: Double, h: Double): MutableList<Double> {
    var points: MutableList<Double> = mutableListOf<Double>()
    var x: Double = a + h
    while (x < (b - h)) {
        points = run { val _tmp = points.toMutableList(); _tmp.add(x); _tmp }
        x = x + h
    }
    return points
}

fun simpson_rule(boundary: MutableList<Double>, steps: Int): Double {
    if (steps <= 0) {
        panic("Number of steps must be greater than zero")
    }
    var a: Double = boundary[0]!!
    var b: Double = boundary[1]!!
    var h: Double = (b - a) / (steps.toDouble())
    var pts: MutableList<Double> = make_points(a, b, h)
    var y: Double = (h / 3.0) * f(a)
    var cnt: Int = (2).toInt()
    var i: Int = (0).toInt()
    while (i < pts.size) {
        var coeff: Double = 4.0 - (2.0 * ((Math.floorMod(cnt, 2)).toDouble()))
        y = y + (((h / 3.0) * coeff) * f(pts[i]!!))
        cnt = cnt + 1
        i = i + 1
    }
    y = y + ((h / 3.0) * f(b))
    return y
}

fun main() {
    run {
        System.gc()
        val _startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _start = _now()
        println(_numToStr(result))
        System.gc()
        val _end = _now()
        val _endMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _durationUs = (_end - _start) / 1000
        val _memDiff = kotlin.math.abs(_endMem - _startMem)
        val _res = mapOf("duration_us" to _durationUs, "memory_bytes" to _memDiff, "name" to "main")
        println(toJson(_res))
    }
}
